### ISSUE TITLE
`newHeads` subscription is missing blocks when a re-org happens

### ISSUE THREAD

**Issue by brianmcgee** [NethermindEth/nethermind#3444]
**Describe the bug**

When subscribed to `newHeads` and a re-org occurs the subscription is supposed to re-publish in order the affected blocks however I have consistently observed that some blocks are missing. For example:

```terminal
13308560 
13308561
13308562
13308561      // re-org begins
13308563      // skipped 13308562
```

**To Reproduce**

You can listen to new heads with the following script, passing as it's argument the file you want to collect block numbers in.

```bash
#!/usr/bin/env bash
set -euo pipefail

echo 'eth_subscribe ["newHeads"]' |
  websocat --jsonrpc --no-close ws://localhost:8546 |
  jq -r --unbuffered 'select(has("params")) | .params.result.number' > "${1}"
```

You can verify the sequence of received block numbers with this script, passing as it's parameter the file in which you are collecting the block numbers.

```bash
#!/usr/bin/env bash
set -euo pipefail

while read -r line; do
  number=$(printf "%d\n" "${line}")
  if [ ! -z ${last_number+x} ]; then
    diff=$((number - last_number))
    echo "$number - diff = $diff"
    if [ "$diff" -le 0 ]; then
      echo "Re-org detected"
    elif [ "$diff" -gt 1 ]; then
      echo "Integrity check failed"
      exit 1
    fi
  fi;

  last_number=${number}
done < "${1}"
```

**Expected behavior**

After listening to new heads for 10 mins or so you should see something like the following after running the verify script

```terminal
13308555 - diff = 1
13308556 - diff = 1
13308557 - diff = 1
13308558 - diff = 1
13308559 - diff = 1
13308560 - diff = 1
13308561 - diff = 1
13308562 - diff = 1
13308561 - diff = -1
Re-org detected
13308563 - diff = 2
Integrity check failed
```

**Comment by sambacha** [comment]
Raw Log Output of our canary if it helps @dceleda - thanks @tkstanczak let us know if you need anything

```txt
11:13:35 INFO   traceId=, parentId=, spanId=, [co.ma.so.Web3Subscriber] (executor-thread-49) New Head = [13313910]
11:13:36 INFO   traceId=, parentId=, spanId=, [co.ma.so.Web3Subscriber] (executor-thread-44) Fetching complete for block=[13313910], elapsed=[1153 ms]
11:13:53 INFO   traceId=, parentId=, spanId=, [co.ma.so.Web3Subscriber] (executor-thread-53) New Head = [13313911]
11:13:53 INFO   traceId=, parentId=, spanId=, [co.ma.so.Web3Subscriber] (executor-thread-26) New Head = [13313910]
11:13:54 INFO   traceId=, parentId=, spanId=, [co.ma.so.Web3Subscriber] (executor-thread-29) Fetching complete for block=[13313911], elapsed=[1189 ms]
11:13:54 WARN   traceId=, parentId=, spanId=, [co.ma.so.Web3Subscriber] (executor-thread-53) Re-publishing blockNumber=[13313910]
11:13:55 INFO   traceId=, parentId=, spanId=, [co.ma.so.Web3Subscriber] (executor-thread-35) Fetching complete for block=[13313910], elapsed=[484 ms]
11:14:07 INFO   traceId=, parentId=, spanId=, [co.ma.so.Web3Subscriber] (executor-thread-50) New Head = [13313912]
11:14:07 ERROR  traceId=, parentId=, spanId=, [co.ma.so.Web3Subscriber] (executor-thread-50) Critical failure: java.lang.Error: Integrity check failed, gap detected: lastPublishedNumber=[13313910], nextNumber=[13313912]
```

**Comment by dceleda** [comment]
I was able to reproduce the issue on Goerli network. I analyzed hashes of parents (parentHash) and it seems that all blocks are published but the order is wrong.
In your case I suspect the correct order should be:

```
13308555 - diff = 1
13308556 - diff = 1
13308557 - diff = 1
13308558 - diff = 1
13308559 - diff = 1
13308560 - diff = 1
13308561 - diff = 1
13308561 - diff = 0  <--
13308562 - diff = 1  <--
13308563 - diff = 1

```

**Comment by dceleda** [comment]
There is a race condition in:
`NewHeadSubscription.OnBlockAddedToMain`
We should queue up notifications and send them in order.

**Comment by dceleda** [comment]
The race condition is there but it's not the main issue. 
The main root cause is we publish in reverse order after a re-org.

**Comment by sambacha** [comment]
Why was this issue closed @LukaszRozmej? Per @dceleda's comment on https://github.com/NethermindEth/nethermind/pull/3458, this is only partially resolved or is this issue superseded by https://github.com/NethermindEth/nethermind/issues/3457?  Just asking so I can track this internally, appreciate the work.

**Comment by dceleda** [comment]
I think it should be closed when #3458 is merged.
